### ISSUE TITLE
Usage decisions

### ISSUE THREAD

**Issue by gwaygenomics** [cytomining/cytominer-transport#1]
How shall users interact with the codebase? Let's track our thoughts and decide here

**Comment by 0x00b1** [comment]
My current thinking for the command line interface:

```
Usage: cytominer-transport [OPTIONS] SOURCE OUTPUT

Options:
  --compartment FILENAME
  --experiments FILENAME
  --images FILENAME
```

**Comment by 0x00b1** [comment]
`compartment` can have multiple invocations, e.g.

```sh
$ cytominer-transport . experiment.parquet \
    --images images.csv \
    --experiments experiments.csv \
    --compartments foo.csv \
    --compartments bar.csv
```

**Comment by shntnu** [comment]
I like it!

tagging @bethac07

**Comment by 0x00b1** [comment]
@gwaygenomics @shntnu @bethac07 Do you want a Python API too (e.g. `to_pandas` and `to_numpy` functions)?

**Comment by gwaygenomics** [comment]
looks good to me.

In cytominer-database, we had an `--engine` option, in which one could specify either `parquet` or `sqlite`. Will cytominer-transport infer this engine from the file extension? I think scrapping sqlite entirely will be good for this package, but it might be worth thinking about splitting an engine option for the inevitability that better data structures become available

Also, if we have noncanonical compartments, would your compartment-naked option be:

```
$ cytominer-transport ./source experiment.parquet --compartment whatever.csv
```

**Comment by 0x00b1** [comment]
> Also, if we have noncanonical compartments, would your compartment-naked option be:

Yes. The user can specify any filename they'd like for `compartment` (`experiment` and `image` too).

**Comment by gwaygenomics** [comment]
> Would you like a Python API (e.g. to_pandas and to_numpy functions in addition to a to_parquet function)?

parquet is 100% compatible with pandas, so no need IMO

**Comment by shntnu** [comment]
I'm very much in favor of scrapping SQLite

> parquet is 100% compatible with pandas, so no need IMO

I agree

**Comment by 0x00b1** [comment]
@gwaygenomics I was thinking something like:

```Python
to_pandas(experiment: Path, image: Path, compartments: List[Path]) -> pandas.DataFrame
```

**Comment by 0x00b1** [comment]
@gwaygenomics and @shntnu Would you still like a public `to_parquet` function?

**Comment by shntnu** [comment]
> Usage: cytominer-transport [OPTIONS] SOURCE OUTPUT

Oh – and this is a big one because `listdir` will not be easy:

`SOURCE` can also be an S3 URL.

**Comment by gwaygenomics** [comment]
> Would you still like a public to_parquet function?

Yeah, I think so. I think our lab will mostly use this package via command line (assay devs final step before handoff), but I can imagine a scenario in which someone would want to run an image-based profiling pipeline end-to-end in python

**Comment by shntnu** [comment]
> Would you still like a public `to_parquet` function?
 think that's wise to have a public API. One use case I can think of: future profiling recipes can use it directly.

**Comment by shntnu** [comment]
> I can imagine a scenario in which someone would want to run an image-based profiling pipeline end-to-end in python

ditto

**Comment by shntnu** [comment]
> `SOURCE` can also be an S3 URL.

Enabling this will likely result in a 5x performance improvement off the bat because our current approach is to [mount the bucket](https://github.com/cytomining/cytominer-vm/blob/78e3ac33279b7e9616017bfeb3746480e126f5dc/s3.sh#L45) using using `s3fs` and then access :D

**Comment by shntnu** [comment]
The following is off-topic (not usage-related) and is getting into implementation, so feel free to ignore for now, or bump to another thread.

It currently takes ~3-5 hours to ingest a file. Our current choice of SQLite does not allow us to do parallel writes, so there's no way to parallelize this.

But we do now, because can store it as a [Parquet Dataset](https://arrow.apache.org/docs/python/parquet.html#partitioned-datasets-multiple-files) which can have multiple files.

So for a 384-well dataset, we can save the output as a Parquet dataset with, say, 24 files (one for each column of the384-well plate). This will also make parallel reads faster, so e.g. aggregation can be faster.

**Comment by 0x00b1** [comment]
@shntnu Good to know. Where does your _n_ (i.e., 24) come from? It can be arbitrary, but ideally, it would correspond to some structural detail of the experiment. Would you mind putting this in a separate issue for tracking purposes?

**Comment by bethac07** [comment]
So is the idea that you would call this on a folder of data, or on one subfolder at a time?  AKA if our structure is the below, is `SOURCE` `plate` or `A01 - Site 1` or either? 
```
-- plate
   |
   -- A01 - Site 1
     | 
     -- Image.csv
     -- Experiment.csv
     -- Foo.csv
   -- A01 - Site 2
     | 
     -- Image.csv
     -- Experiment.csv
     -- Foo.csv
```

**Comment by bethac07** [comment]
(I'm fine with either behavior, as long as we're all on the same page as to what it is; in cytominer-database, we'd be calling it on `plate`, but my reading of the suggested usage would be that going forward we'd be calling it on each site, but those dumping back into a common `experiment.parquet`

**Comment by shntnu** [comment]
> So is the idea that you would call this on a folder of data, or on one subfolder at a time?

Not sure if the q is for @0x00b1 but I think we should call it on a folder of data; `SOURCE` is `plate`.

I have a related implementation comment here https://github.com/cytomining/cytominer-transport/issues/2#issuecomment-812663797
